### ISSUE TITLE
pytorch cpp api the derivative for 'target' is not implemented

### ISSUE THREAD

**Issue by mhubii** [pytorch/pytorch#16830]
Code to reproduce the error. Is it a bug on my machine?

Main
```cpp
#include <torch/torch.h>
#include <iostream>

int main() {

	// Layer.
	torch::nn::Linear linear1(5, 1);
	torch::nn::Linear linear2(5, 1);

	// Optimizer.
	torch::optim::Adam opt(linear1->parameters(), torch::optim::AdamOptions(0.001));

	// Input.
	torch::Tensor in = torch::randn({1, 5});

	// Output. Have one of these tensors. The first one wont work.
	torch::Tensor desired_out = linear2->forward(in); // raises the error "the derivative for 'target' is not implemented"
	//torch::Tensor desired_out = torch::ones({1, 1}); // works perfectly fine

	// Training loop.
	std::cout << "initial: " << linear1->forward(in) << std::endl;

	for (int i = 0; i < 1000; i++) {

		opt.zero_grad();
		torch::Tensor out = linear1->forward(in);
		torch::Tensor loss = torch::mse_loss(out, desired_out);
		loss.backward();
		opt.step();	
	}

	std::cout << "desired: " << desired_out << std::endl;
	std::cout << "trained: " << linear1->forward(in) << std::endl;

	return 0;
}
```

CMake
```cmake
cmake_minimum_required(VERSION 3.11 FATAL_ERROR)

find_package(Torch REQUIRED)

include_directories(${Torch_INCLUDE_DIRS})

add_executable(main main.cpp)
target_link_libraries(main ${TORCH_LIBRARIES})
```

The Python equivalent of this code can be fixed by retaining the graph. May this be my mistake. How can I keep the graph in the backward pass?

**Comment by mhubii** [comment]
In the Pytorch forum, we solved the question. Still, is it a bug? See

[https://discuss.pytorch.org/t/backward-keep-graph-cpp-api/36559](url)

**Comment by gchanan** [comment]
CC @yf225.

**Comment by fmassa** [comment]
in Python we have a separate dispatch for losses in some cases, like when the target requires gradient https://github.com/pytorch/pytorch/blob/72f070a124c3062939873a5468463d7c9e3ac4eb/torch/nn/functional.py#L2177-L2180

the easy solution is to just do the same thing in C++ (implement those cases with PyTorch ops).

**Comment by timha92** [comment]
Try add detach() to your tensor 
either as
torch::Tensor loss = torch::mse_loss(out, desired_out.detach());
or 
torch::Tensor desired_out = linear2->forward(in).detach();

**Comment by mhubii** [comment]
yes thank you @timha92, that is the solution which was also provided in the [forum](https://discuss.pytorch.org/t/backward-keep-graph-cpp-api/36559/2) :)

**Comment by soumith** [comment]
it's still a bug, that we should fix. I believe we shouldn't have to `detach` explicitly, because we dont need to do that in python

**Comment by yf225** [comment]
I believe we will need to add the separate dispatch for `mse_loss` in C++ API for the case when `target` requires gradient, as @fmassa mentioned. I will work on it.

**Comment by yf225** [comment]
@glaringlee please feel free to prioritize it :D

**Comment by glaringlee** [comment]
This should be fixed in #44437
